Add a script to fetch and diff GitHub workflow logs

krcroft opened the pull request:
Adds a stand-alone script `./scripts/diff-workflow-against-master.sh` that does a couple things:
1. Pulls the artifacts and logs for the local branch's latest CI run
2. Diffs any failed jobs with the most recent successful equivalent master branch job

The script is fire-and-forget; it takes no argument and just grinds away for a couple seconds until done.  Even if you're not interested in diffing, it's a convenient way to slurp down our current zip files, reports, and logs. 

It currently constructs a directory in `/tmp` to hold content, and the script will clean away older files too so it doesn't get overrun.

The log diff needs improving though.  @dreamer , if you have something smarter than can take in two "raw logs" and diff-out the meaningful gcc differences (and drop everthing elsee), then we can drop that function (or python) in to improve it.  Right now I'm just stripping the prefix time-stamp and then diffing the raw logs; so the resulting diff file is still pretty cluttered.

None the less, it's a first cut that's working - and we can refine it as we go.  Let me know what you think!

Here's output from the start of the run:

![2020-02-11_23-57](https://user-images.githubusercontent.com/1557255/74322456-cfeacb80-4d38-11ea-8c32-3b71a3b318d7.png)

krcroft commented:
The script will step you through how to generate a personal auth token.  

I _think_ anyone with a GitHub account should be able do this - so anyone curious out there, please kick the tires and report back!

dreamer commented on scripts/diff-workflow-against-master.sh:
```diff
@@ -0,0 +1,471 @@
+	local search_key="$2"
+	local search_value="$3"
+	local return_keys="$4"
+	jq -r --arg value "$search_value"\
```
`./scripts/diff-workflow-against-master.sh: line 274: jq: command not found`

Add a dependency check on script start; if a user has no `jq` installed, the script should immediately exit with an error (before a user has a chance to generating a token).

dreamer replied:
I know that token is saved before this happens, but some user might get lost and generate a new one.

krcroft replied:
Huge catch here; I'm baffled why I didn't do a requirements check first - thank you @dreamer !
Will run a check for `unzip` and any other non-core-util commands I might be using.

krcroft replied:
Added (for all tools)

dreamer commented on scripts/diff-workflow-against-master.sh:
```diff
@@ -0,0 +1,471 @@
+		return
+	fi
+	# Get the username
+	clear
```
Don't clear screen, please.

krcroft replied:
It felt a bit more instructive, but I'm 50-50 on how it turned out. A real  TUI would have been cool with something like `dialog` :-)  But yeah, will drop the `clear`s!

krcroft replied:
Done!

dreamer commented on scripts/diff-workflow-against-master.sh:
```diff
@@ -0,0 +1,471 @@
+	done # workflows
+}
+
+main
```
Use `main "$@"`; we don't handle any parameters yet, but let's prepare for it.

krcroft replied:
Makes sense; will do.

dreamer commented on scripts/diff-workflow-against-master.sh:
```diff
@@ -0,0 +1,471 @@
+	declare -gr cache_dir
+	declare -gr diffs_dir
+	declare -gr logs_dir
+	echo "Initializing storage area: $parent"
```
It took me a while to figure out where the logs were actually downloaded (I missed this line when reading output) - can you remove this echo from here, place it just before program exits (so it'll be the last line of output) and give make the message less technical, e.g.: "Copy of logs in: <path>".

krcroft replied:
Will do; much more convenient!

krcroft replied:
Added.

krcroft commented:
> store token in the repo in .git/config file instead of ~/.netrc

The token is equivalent to a password, which means its storage point should be secured by default. Git's tools create `.git/config` with default system permissions (typically 644):
```
[kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll .git/config 
-rw-r--r-- 1 kcroft users 1.7K Feb 12 08:06 .git/config
```
Unlike SSH's tools, that create sensitive files with restrictive permissions:
```
[kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll -d ~/.ssh ~/.ssh/id_ed25519
drwx------ 2 kcroft users 4.0K Nov 28 10:35 /home/kcroft/.ssh
-rw------- 1 kcroft users  411 Aug  4  2017 /home/kcroft/.ssh/id_ed25519
```

Similarly, `netrc` is dedicated to storing credentials and is is the recommended place to store sensitive Curl credentials: https://community.apigee.com/articles/39911/do-you-use-curl-stop-using-u-please-use-curl-n-and.html

We could make the script lock down the `.git` directory and its `config` file -- but that doesn't solve external assumption that `.git/config` is just a regular repo file.  

So some users might zip-up their repo and send it to colleagues; or might be backed up into an IT-manager server farm dedicated to GIT repositories having lower security access requirements than where they back-up secret files.. (mentioning this because some companies give workers some hours/week for personal projects). 

> pre-fill email field with git config user.email

For users who have enabled "Keep my email addresses private" in their user settings, then their `config` will be set with the no-reply address as well, such as `email = krcroft@users.noreply.github.com` (and that won't work when generating a token). But I can certainly try to pull it from here first :), and if it happens to be a noreply address then fallback to asking the user.

dreamer commented:
> > store token in the repo in .git/config file instead of ~/.netrc
> 
> The token is equivalent to a password, which means its storage point should be secured by default. Git's tools create `.git/config` with default system permissions (typically 644):
> 
> ```
> [kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll .git/config 
> -rw-r--r-- 1 kcroft users 1.7K Feb 12 08:06 .git/config
> ```
> 
> Unlike SSH's tools, that create sensitive files with restrictive permissions:

`.git/config` is just a user's file, we can safely change mode on it; but I wo't press the issue (I just dislike placing new config files directly in my home dir).

> ```
> [kcroft@hoya dosbox-staging] (kc/issue-diff-1 $) ll -d ~/.ssh ~/.ssh/id_ed25519
> drwx------ 2 kcroft users 4.0K Nov 28 10:35 /home/kcroft/.ssh
> -rw------- 1 kcroft users  411 Aug  4  2017 /home/kcroft/.ssh/id_ed25519
> ```
> 
> Similarly, `netrc` is dedicated to storing credentials and is is the recommended place to store sensitive Curl credentials: https://community.apigee.com/articles/39911/do-you-use-curl-stop-using-u-please-use-curl-n-and.html
> 
> We could make the script lock down the `.git` directory and its `config` file -- but that doesn't solve external assumption that `.git/config` is just a regular repo file.
> 
> So some users might zip-up their repo and send it to colleagues; or might be backed up into an IT-manager server farm dedicated to GIT repositories having lower security access requirements than where they back-up secret files.. (mentioning this because some companies give workers some hours/week for personal projects).

Yeah, that would be bad… Users are not supposed to do that (receiving zipped git repo is not secure for receiver), but some people might try it.

> > pre-fill email field with git config user.email
> 
> For users who have enabled "**Keep my email addresses private**" in their user settings, then their `config` will be set with the no-reply address as well, such as `email = krcroft@users.noreply.github.com` (and that won't work when generating a token). But I can certainly try to pull it from here first :), and if it happens to be a noreply address then fallback to asking the user.

You misunderstood my suggestion: I think we can safely pre-fill input to email (not skip it altogether), so when it matches user can just press enter.

`read -r -p "mail: " -e -i "foo@bar.baz" variable_name`

mail address will be placed as suggested value, user can press enter or edit it :)

krcroft commented:
Just came across https://git-scm.com/docs/git-credential-store probably the best option; then use an environment variable to pass the token to curl, so it doesn't show up on the command line.

dreamer commented:
> Just came across https://git-scm.com/docs/git-credential-store probably the best option; then use an environment variable to pass the token to curl, so it doesn't show up on the command line.

Nice! I didn't know it exists :) (I don't feel we need to implement it during this PR - your call).

krcroft commented:
All set for another round:
 - dependencies (like `jq`) checked
 - switched to git credentials
 - pre-populate email, but only if it's a non-noreply address
 - print storage directory at the end
 - args passed to main
 - removed the `clear`s during setup
 - a bit more verbose when `ctrl-c` is pressed

krcroft commented:
Suggestion for round two PR: trial your last suggestions: 

>  easy way to create diffs - at least for compilation logs would be to run count-warnings.py -lf <file> on specific logs and then run diff on outputs of warning counter; instead of diff we could even run merge.tool/merge.guitool automatically (description of these options in man git-config).

dreamer commented:
GitHub interface got confused after I merged this change in manually. It's now available on master branch 675d8b99bf5148d406c591b630c64861cb02d466

@krcroft Thank you! :)
